# CRUD in MongoDB:

In this notebook we will explore the main transactions done on a database (CRUD) on a document based database (MongoDB).

First, we load the data from the file primer-dataset.json using cmd using the following command: 

mongoimport --db test --collection restaurants --drop --file "path to file"

In [296]:
import pymongo as pg #Among other things handles conncection between the db and python environement 
from pprint import pprint #pretty print
import datetime #Needed to format the datetime input later on.

In [297]:
client=pg.MongoClient()
db=client.test

In [298]:
print(db.restaurants.find_one())

{'_id': ObjectId('5cc788443ff99b33afc973b5'), 'address': {'building': '351', 'coord': [-73.98513559999999, 40.7676919], 'street': 'West   57 Street', 'zipcode': '10019'}, 'borough': 'Manhattan', 'cuisine': 'Irish', 'grades': [{'date': datetime.datetime(2014, 9, 6, 0, 0), 'grade': 'A', 'score': 2}, {'date': datetime.datetime(2013, 7, 22, 0, 0), 'grade': 'A', 'score': 11}, {'date': datetime.datetime(2012, 7, 31, 0, 0), 'grade': 'A', 'score': 12}, {'date': datetime.datetime(2011, 12, 29, 0, 0), 'grade': 'A', 'score': 12}], 'name': 'Dj Reynolds Pub And Restaurant', 'restaurant_id': '30191841'}


In [299]:
pprint(db.restaurants.find_one())

{'_id': ObjectId('5cc788443ff99b33afc973b5'),
 'address': {'building': '351',
             'coord': [-73.98513559999999, 40.7676919],
             'street': 'West   57 Street',
             'zipcode': '10019'},
 'borough': 'Manhattan',
 'cuisine': 'Irish',
 'grades': [{'date': datetime.datetime(2014, 9, 6, 0, 0),
             'grade': 'A',
             'score': 2},
            {'date': datetime.datetime(2013, 7, 22, 0, 0),
             'grade': 'A',
             'score': 11},
            {'date': datetime.datetime(2012, 7, 31, 0, 0),
             'grade': 'A',
             'score': 12},
            {'date': datetime.datetime(2011, 12, 29, 0, 0),
             'grade': 'A',
             'score': 12}],
 'name': 'Dj Reynolds Pub And Restaurant',
 'restaurant_id': '30191841'}


In [300]:
db.restaurants.count() #Number of restaurants in the JSON file.

C:\Users\YsfEss\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: DeprecationWarning: count is deprecated. Use estimated_document_count or count_documents instead. Please note that $where must be replaced by $expr, $near must be replaced by $geoWithin with $center, and $nearSphere must be replaced by $geoWithin with $centerSphere
  """Entry point for launching an IPython kernel.


25359

# Inserting a new document:

In [208]:
db.restaurants.insert_one(
{
"address" : {
"street" : "2 Avenue",
"zipcode" : "10075",
"building" : "1480",
"coord" : [ -73.9557413, 40.7720266 ]
},
"borough" : "Manhattan",
"cuisine" : "Italian",
"grades" : [
{
"date" : datetime.datetime(2014, 10, 1, 0, 0, 0, 0),
    "grade" : "A",
"score" : 11
},
{
"date" : datetime.datetime(2014, 1, 16, 0, 0, 0, 0),
"grade" : "B",
"score" : 17
}
],
"name" : "Vella",
"restaurant_id" : "41704620"
}
)

In [209]:
db.restaurants.count()

C:\Users\YsfEss\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: DeprecationWarning: count is deprecated. Use estimated_document_count or count_documents instead. Please note that $where must be replaced by $expr, $near must be replaced by $geoWithin with $center, and $nearSphere must be replaced by $geoWithin with $centerSphere
  """Entry point for launching an IPython kernel.


25360

# Updating Top-level fields:

Update the Document relative to the Juni restaurant and set the cuisine to ‘American
(New)’ and the lastModified field in current date to True. Use the function find_one() before and after the update to
verify its execution.

In [316]:
queryF = { "name": "Juni" }
pprint(db.restaurants.find_one(queryF))
newValues={"$set":{"cuisine":"American(New)"}}
db.restaurants.update_one(queryF,newValues)
pprint(db.restaurants.find_one(queryF))

{'_id': ObjectId('5cc788443ff99b33afc98b45'),
 'address': {'building': '12',
             'coord': [-73.9852329, 40.745971],
             'street': 'East 31 Street',
             'zipcode': '10016'},
 'borough': 'Manhattan',
 'cuisine': 'American',
 'grades': [{'date': datetime.datetime(2014, 9, 19, 0, 0),
             'grade': 'A',
             'score': 12},
            {'date': datetime.datetime(2013, 8, 5, 0, 0),
             'grade': 'A',
             'score': 5},
            {'date': datetime.datetime(2012, 6, 7, 0, 0),
             'grade': 'A',
             'score': 0}],
 'name': 'Juni',
 'restaurant_id': '41156888'}
{'_id': ObjectId('5cc788443ff99b33afc98b45'),
 'address': {'building': '12',
             'coord': [-73.9852329, 40.745971],
             'street': 'East 31 Street',
             'zipcode': '10016'},
 'borough': 'Manhattan',
 'cuisine': 'American(New)',
 'grades': [{'date': datetime.datetime(2014, 9, 19, 0, 0),
             'grade': 'A',
             'score': 12},
 

# Updating Embedded fields:

Update the street of the restaurant whose id is 41156888 with the new value "East
41st Street"

In [130]:
query2 = { "restaurant_id": "41156888" }
pprint(db.restaurants.find_one(query2))
newValues={"$set":{"address.street":"East 41st Street"}}
db.restaurants.update_one(query2,newValues)
pprint(db.restaurants.find_one(query2))

{'_id': ObjectId('5cc731653ff99b33afc84a98'),
 'address': {'building': '12',
             'coord': [-73.9852329, 40.745971],
             'street': 'East 31 Street',
             'zipcode': '10016'},
 'borough': 'Manhattan',
 'cuisine': 'American(New)',
 'grades': [{'date': datetime.datetime(2014, 9, 19, 0, 0),
             'grade': 'A',
             'score': 12},
            {'date': datetime.datetime(2013, 8, 5, 0, 0),
             'grade': 'A',
             'score': 5},
            {'date': datetime.datetime(2012, 6, 7, 0, 0),
             'grade': 'A',
             'score': 0}],
 'name': 'Juni',
 'restaurant_id': '41156888'}
{'_id': ObjectId('5cc731653ff99b33afc84a98'),
 'address': {'building': '12',
             'coord': [-73.9852329, 40.745971],
             'street': 'East 41st Street',
             'zipcode': '10016'},
 'borough': 'Manhattan',
 'cuisine': 'American(New)',
 'grades': [{'date': datetime.datetime(2014, 9, 19, 0, 0),
             'grade': 'A',
             'score':

# Updating multiple Documents:

Update all documents that have a zipcode field equal to "10016" and cuisine
field equal to “Other”. Set the cuisine field to "Category To Be Determined".

There is no field called "lastModified".

In [131]:
db.restaurants.find({"lastModified": {"$exists":True}}).count()

C:\Users\YsfEss\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: DeprecationWarning: count is deprecated. Use Collection.count_documents instead.
  """Entry point for launching an IPython kernel.


0

In [132]:
db.restaurants.find({"cuisine":"Other"}).count() #Number of restaurants with 'Other' value for the cuisine key.

C:\Users\YsfEss\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: DeprecationWarning: count is deprecated. Use Collection.count_documents instead.
  """Entry point for launching an IPython kernel.


1011

In [133]:
db.restaurants.update_many({"cuisine":"Other"},{"$set":{"cuisine":"Category To Be Determined"}}) #Updating multiple files.

In [134]:
db.restaurants.find({"cuisine":"Other"}).count() #We expect 0.

C:\Users\YsfEss\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: DeprecationWarning: count is deprecated. Use Collection.count_documents instead.
  """Entry point for launching an IPython kernel.


0

In [135]:
db.restaurants.find({"cuisine":"Category To Be Determined"}).count() # We expect 1011

C:\Users\YsfEss\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: DeprecationWarning: count is deprecated. Use Collection.count_documents instead.
  """Entry point for launching an IPython kernel.


1011

# Replacing a Document:

Update the Document for the restaurant with id : “41704620”. The modified document will only contain the _id field, name field, the address field. i.e. the document will not contain the restaurant_id, cuisine, grades, and the borough fields.

In [210]:
name=db.restaurants.find_one({"restaurant_id":"41704620"},{"name"})
address=db.restaurants.find_one({"restaurant_id":"41704620"},{"address"})

In [211]:
def extractSubFields(field):
    l=[]
    for subField in field.keys():
        l.append(field[subField])
    return(l)

In [216]:
db.restaurants.find_one({"restaurant_id":"41704620"})

{'_id': ObjectId('5cc75cf23ff99b33afc94b39'),
 'address': {'building': '1480',
  'coord': [-73.9557413, 40.7720266],
  'street': '2 Avenue',
  'zipcode': '10075'},
 'borough': 'Manhattan',
 'cuisine': 'Italian',
 'grades': [{'date': datetime.datetime(2014, 10, 1, 0, 0),
   'grade': 'A',
   'score': 11},
  {'date': datetime.datetime(2014, 1, 16, 0, 0), 'grade': 'B', 'score': 17}],
 'name': 'Vella',
 'restaurant_id': '41704620'}

In [218]:
db.restaurants.replace_one({"restaurant_id":"41704620"},{
    "_id":extractSubFields(name)[0],
    "address":extractSubFields(address)[1],
    "name":extractSubFields(name)[1]
})

In [219]:
db.restaurants.find_one({"name":"Vella"})

{'_id': ObjectId('5cc75cf23ff99b33afc94b39'),
 'address': {'building': '1480',
  'coord': [-73.9557413, 40.7720266],
  'street': '2 Avenue',
  'zipcode': '10075'},
 'name': 'Vella'}

# Querying Data:

Find all restaurants that were last modified this week.

In [327]:
# Handeling datetime type.
ObjectId('5cc788443ff99b33afc98b45')
lastMod=ObjectId('5cc788443ff99b33afc98b45').generation_time.replace(tzinfo=None) #removing timezone to be able to use operation on it.
now=datetime.datetime.now()
d=now-lastMod # its type is datetime.timedelta. datetime.timedelta([days,] [seconds,] [microseconds,] [milliseconds,] [minutes,] [hours,] [weeks])
# We can access each component of d by its name for example number of days between the two dates d.days.
d

AttributeError: can't set attribute

In [324]:
now=datetime.datetime.now()
l=[]
for doc in db.restaurants.find():
    dif=ObjectId(doc["_id"]).generation_time.replace(tzinfo=None)-now
    if dif.seconds<:
        l.append(doc["_id"])
lastModifiedDocs=db.restaurants.find({"_id":{"$in":l}})

In [325]:
l[0:10]

[ObjectId('5cc788443ff99b33afc973b5'),
 ObjectId('5cc788443ff99b33afc973b6'),
 ObjectId('5cc788443ff99b33afc973b7'),
 ObjectId('5cc788443ff99b33afc973b8'),
 ObjectId('5cc788443ff99b33afc973b9'),
 ObjectId('5cc788443ff99b33afc973ba'),
 ObjectId('5cc788443ff99b33afc973bb'),
 ObjectId('5cc788443ff99b33afc973bc'),
 ObjectId('5cc788443ff99b33afc973bd'),
 ObjectId('5cc788443ff99b33afc973be')]

In [326]:
lastModifiedDocs.count()

C:\Users\YsfEss\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: DeprecationWarning: count is deprecated. Use Collection.count_documents instead.
  """Entry point for launching an IPython kernel.


25359

In [315]:
for doc in db.restaurants.find():
    print((ObjectId(doc["_id"]).generation_time.replace(tzinfo=None)-now).seconds)

82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
8267

82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
8267

82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
8267

82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
8267

82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
8267

82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
8267

82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
82671
8267

82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
8267

82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
8267

82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
8267

82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
8267

82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
8267

82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
8267

82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
8267

82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
8267

82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
8267

82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
82672
8267

In [270]:
b

datetime.datetime(2019, 4, 29, 23, 15, 46, 260679)

In [268]:
a

datetime.datetime(2019, 4, 29, 20, 22, 10)

In [271]:
b-a

datetime.timedelta(0, 10416, 260679)